src/main/java/org/passay/dictionary/MemoryMappedFileWordList.java

```diff
@@ -0,0 +1,176 @@
+   * called subsequently.</p>
+   *
+   * @param  raf  File containing words, one per line.
+   * @param  caseSensitive  Set to true to create case-sensitive word list, false otherwise.
+   * @param  cachePercent  Percent (0-100) of file to cache in memory for improved read performance.
+   *
+   * @throws  IllegalArgumentException  if cache percent is out of range.
+   * @throws  IOException  if an error occurs reading the supplied file
+   */
+  public MemoryMappedFileWordList(final RandomAccessFile raf, final boolean caseSensitive, final int cachePercent)
+    throws IOException
+  {
+    super(raf, caseSensitive, cachePercent);
+    final FileChannel channel = file.getChannel();
+    buffer = channel.map(FileChannel.MapMode.READ_ONLY, 0, channel.size());
+
+    final int fileBytes = buffer.capacity();
+    final int cacheSize = (fileBytes / HUNDRED_PERCENT) * cachePercent;
+    final int cacheOffset = cacheSize == 0 ? fileBytes : cacheSize > fileBytes ? 1 : fileBytes / cacheSize;
+
+    long pos = 0;
+    String a;
+    String b = null;
+    while (buffer.hasRemaining()) {
+      a = readLine(buffer);
+      if (b != null && comparator.compare(a, b) < 0) {
+        throw new IllegalArgumentException("File is not sorted correctly for this comparator");
+      }
+      b = a;
+
+      if (cacheSize > 0 && size % cacheOffset == 0) {
+        cache.put(size, pos);
+      }
+      pos = buffer.position();
+      size++;
+    }
+  }
+
+
+  /**
+   * Reads the file line by line and returns the word at the supplied index. Returns null if the index cannot be read.
+   * This method leverages the cache to seek to the closest position of the supplied index.
+   *
+   * @param  index  to read word at
+   *
+   * @return  word at the supplied index
+   *
+   * @throws  IllegalStateException  if an error occurs reading the supplied file
+   */
+  @Override
+  protected String readFile(final int index)
+  {
+    synchronized (buffer) {
+      int i = 0;
+      if (!cache.isEmpty() && cache.firstKey() <= index) {
+        i = cache.floorKey(index);
+      }
+
+      final int pos = i > 0 ? cache.get(i).intValue() : 0;
+      buffer.position(pos);
+
+      String s;
+      while (buffer.hasRemaining()) {
+        s = readLine(buffer);
+        if (i == index) {
+          return s;
+        }
+        i++;
+      }
+    }
+    return null;
+  }
+
+
+  /**
+   * Reads a line from the supplied buffer. Buffer position will be set to the beginning of the next line. Line
+   * terminator must be one of '\n', '\r', or "\r\n".
+   *
+   * @param  buffer  to read from
+   *
+   * @return  file line
+   */
+  private static String readLine(final MappedByteBuffer buffer)
+  {
+    final StringBuilder line = new StringBuilder();
+    boolean lineFeed = false;
+    boolean carriageReturn = false;
+    while (buffer.hasRemaining()) {
+      final char c = (char) buffer.get();
+      if (lineFeed) {
```

Comment:
This implementation struck me as overly complicated. I took the time to implement a simpler routine that I manually tested over freebsd.sort and a CRLF version of the same on my local machine. I'm going to take the liberty to commit that change onto this branch; you can veto it if you don't like it for some reason.